raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1914,8 +1915,7 @@ private long pollLeader(long currentTimeMs) {
         LeaderState<T> state = quorum.leaderStateOrThrow();
         maybeFireLeaderChange(state);
 
-        GracefulShutdown shutdown = this.shutdown.get();
-        if (shutdown != null) {
+        if (shutdown.get() != null || state.isResignRequested()) {
```

Comment:
unrelated, but maybe worth creating helper method that returns `Optional<GracefulShutdown>` to avoid these null checks throughout

Reply:
That's fair. Let me see if it's reasonable to do this here or if we should push to a separate PR.

Reply:
If you don't mind, let's do this refactor separately. There are a fair number of uses that would benefit from having `Optional<GracefulShutdown>`.

Reply:
Works for me 👍